doc/cookbook/using-free-client/UsingFreeClient.lhs

```diff
@@ -0,0 +1,173 @@
+# Using Free Client (for tests)
+
+Someone asked on IRC about getting `Request` & `Response` of what
+`servant-client` uses, for testing purposes.  My response: to ad-hoc extend
+`servant-client` (as for tests), use `Servant.Client.Free`. This recipe is an
+evidence.
+
+First the module header, but this time We'll comment the imports.
+
+```haskell
+{-# LANGUAGE DataKinds     #-}
+{-# LANGUAGE TypeOperators #-}
+module Main (main) where
+```
+
+We will primarily use `Servant.Client.Free`, it doesn't re-export anything
+from `free` package, so we need to import it as well.
+
+```haskell
+import Control.Monad.Free
+import Servant.Client.Free
+```
+
+Also we'll use `servant-client` internals, which uses `http-client`,
+so let's import them *qualified*
+
+```haskell
+import qualified Servant.Client.Internal.HttpClient as I
+import qualified Network.HTTP.Client                as HTTP
+```
+
+The rest of the imports are for a server we implement here for completeness.
+
+```haskell
+import Servant
+import Network.Wai.Handler.Warp (run)
+import System.Environment       (getArgs)
+```
+
+## API & Main
+
+We'll work with a very simple API:
+
+```haskell
+type API = "square" :> Capture "n" Int :> Get '[JSON] Int
+
+api :: Proxy API
+api = Proxy
+```
+
+Next we implement a `main`. If passed `server` it will run `server`, if
+`client` a small `test` (to be defined next) will be run. This should be pretty
+straigh-forward:
+
+```haskell
+main :: IO ()
+main = do
+    args <- getArgs
+    case args of
+        ("server":_) -> do
+            putStrLn "Starting cookbook-using-free-client at http://localhost:8000"
+            run 8000 $ serve api $ \n -> return (n * n)
+        ("client":_) ->
+            test
+        _ -> do
+            putStrLn "Try:"
+            putStrLn "cabal new-run cookbook-using-free-client server"
+            putStrLn "cabal new-run cookbook-using-free-client client"
+```
+
+## Test
+
+In the actual test, we'll use a `Servant.Client.Free` client.
+Cecause we have a single endpoint API, we'll get a single client function:
+
+```haskell
+fcli :: Int -> Free ClientF Int
+fcli = client api
+```
+
+Next, we can write our small test. We'll pass a value to `fcli` and inspect
+the `Free` structure. First three possibilities are self-explanatory:
+
+```haskell
+test :: IO ()
+test = case fcli 42 of
+    Pure n ->
+        putStrLn $ "ERROR: got pure result: " ++ show n
+    Free (Throw err) ->
+        putStrLn $ "ERROR: got error right away: " ++ show err
+    Free (StreamingRequest _req _k) ->
+        putStrLn $ "ERROR: need to do streaming request" -- TODO: no Show Req :(
+```
+
+We are interested in `RunRequest`, that's what client should block on:
```

Comment:
I'd suggest: "that's the only http client backend operation used by a client for a simple API like the one above", or something like that.